### DISCUSSION TITLE
Looking for a sample code to create an issue resolution service issue

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#2831] [Q&A]
I'm not having much luck generating a POST request to create an issue in IRS that doesn't return an error.  (I'm able to make GET requests using code from the Issue Sample but this sample doesn't demonstrate POST.)

Can someone provide a code snippet I can use to get started?

Thanks,

**Comment by aruniverse** [comment]
Have you taken a look at the [docs](https://developer.bentley.com/apis/issues-v1/operations/create-issue/)? There's a _"Try it out"_ button you can use to test. Not sure if we have any samples for creation, I'll check and update if there is.

Does your client have the `issues:modify` scope and the `Forms_CreateAccess` permission for the project?

**Comment by JohnMajerle** [comment]
Hello Arun,

Yes to both questions.  I would never post a question before reading the documentation first.  

Below is the **_createIssue_** method I added to **_IssueClient.ts_**  When the REST web request is invoked here or from **_Try it out_** I get the same error message:

```
IssueClient.create() error = Error: {
  "error": {
    "code": "UnknownError",
    "message": "An unexpected error occurred while making a request to the backend server."
  }
}
```
Unfortunately this message is not particularly useful for diagnosing the problem.

Note that I didn't set all the possible values in the body.  I had  assumed these were optional but perhaps they are not.

I am able to create  issues  using the **_Design Review_** application but unfortunately that form is not open source so I can't look at it to get clues as to what might be going wrong in my code.  

Thanks,

--John M.



```
  public static async createIssue (elementId: string, subject: string, description: string, dueDate: string, 
                                   status: string, assigneeDisplayName: string, assigneeId: string) {

    const issue = {
      "subject": subject,
      "description": description,
      "formId": "hVqQ2d4N40O7taUCpHB8nZtNR9lXgvpPnmDGp3zz_HI",
      "dueDate": dueDate,
      "status": status,
      "assignee": {
          "displayName": assigneeDisplayName,
          "id": assigneeId
      },
      "assignees": [],
      "sourceEntity": {
          "iModelElement": {
              "modelName": UiFramework.getIModelConnection()!.name,
              "modelId": UiFramework.getIModelConnection()!.iModelId!,
              "elementId": elementId,
              "changeSetId": UiFramework.getIModelConnection()!.changeset.id
          }
      },
    }

    const url = `https://api.bentley.com/issues/`;

    const options : RequestOptions = {
      method: "POST",
      headers: {
        Authorization: (await IssuesClient.getAccessToken())?.toTokenString(IncludePrefix.Yes),
        contentType: "application/json",
        Accept: "application/vnd.bentley.itwin-platform.v1+json"
      },
      body: issue,
    };

    return request(await IssuesClient.getRequestContext(), url, options)
              .then((resp: Response): AuditTrail | undefined => {
                if (resp.body === undefined) {
                  console.error (`IssueClient.create() resp.body = undefined`);
                  return undefined;                  
                }

                console.log (`IssueClient.create() resp.body = ${resp.body}`);
                return resp.body;
              }).catch((_reason: any) => {

                console.error (`IssueClient.create() error = ${_reason}`);
                return undefined;
              });
  }
```

**Reply by aruniverse** [reply]
_"I would never post a question before reading the documentation first."_ Bless you, John :) Just wanted to make sure to cover all bases.

I'm asking someone more familiar with the Issues api and we're looking into the error you're getting. In the failing request on your end, there should be a header with `x-correlation-id`, could you share that with us?

**Reply by ajmiam** [reply]
Hi, John.

I took a look at the Seq logs and found the error the backend is returning:
`Workflow doesn't support starting with status open`
(I apologize for how unclear this error was; we can work on returning more useful details for this error case in the future.)

This error means that the value of the "status" property of the issue body you are writing is invalid according to the workflow for the issue's type (or the issue's type doesn't have a workflow, in which case you can omit the status property).

This tutorial might help:
https://developer.bentley.com/tutorials/stepping-through-an-issue-workflow/

You'll need to use this endpoint to get the workflow for the issue you want to create.  If one exists, the `startStates` array will tell you the possible values of the `status` property when you're creating a new issue from that form definition.
https://developer.bentley.com/apis/issues-v1/operations/get-workflow/

The issue's type is determined from the form definition.  Make sure the "type" parameter in the URL for the "get workflow" request matches the "type" property returned when you use this operation to GET the form definition whose ID you set as "formId" in the issue body:
https://developer.bentley.com/apis/issues-v1/operations/get-form-definition-by-id/

Note:  If you get a 404 error from the get-workflow endpoint, it means there is no workflow for that type--which is an allowed condition, and means you should not set a status when creating the issue.

Please let me know if you have further questions!

**Reply by JohnMajerle** [reply]
Thanks for those links.  They were very helpful.
It's unfortunate the service SEQ logs are Bentley internal.  They can be incredibly helpful diagnosing what is going wrong on the service side.  Barring that, useful service error messages would be a big help.  An error code of "**_UnknownError_**" should never be returned from  a REST call.
